### PR TITLE
TST: Fixed timezone issues post DatetimeArray refactor

### LINKED ISSUE #18338: BUG: Unstacking MultiIndex with datetime tz-aware data raises ValueError: cannot create a DatetimeTZBlock without a tz
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({
    'timestamp': [pd.Timestamp('2017-08-27 01:00:00.709949+0000', tz='UTC')],
    'a': ['a'],
    'b': ['b'],
    'c': ['c'],
})
df.set_index(['a', 'b']).unstack()
```
#### Problem description
Exception thrown. This used to work pre 0.21 and works with non Timestamp columns
```
ValueError                                Traceback (most recent call last)
<ipython-input-13-469280bd9dc1> in <module>()
      5     'c': ['c'],
      6 })
----> 7 df.set_index(['a', 'b']).unstack()

/usr/local/lib/python3.5/dist-packages/pandas/core/frame.py in unstack(self, level, fill_value)
   4544         """
   4545         from pandas.core.reshape.reshape import unstack
-> 4546         return unstack(self, level, fill_value)
   4547 
   4548     _shared_docs['melt'] = ("""

/usr/local/lib/python3.5/dist-packages/pandas/core/reshape/reshape.py in unstack(obj, level, fill_value)
    467     if isinstance(obj, DataFrame):
    468         if isinstance(obj.index, MultiIndex):
--> 469             return _unstack_frame(obj, level, fill_value=fill_value)
    470         else:
    471             return obj.T.stack(dropna=False)

/usr/local/lib/python3.5/dist-packages/pandas/core/reshape/reshape.py in _unstack_frame(obj, level, fill_value)
    480         unstacker = partial(_Unstacker, index=obj.index,
    481                             level=level, fill_value=fill_value)
--> 482         blocks = obj._data.unstack(unstacker)
    483         klass = type(obj)
    484         return klass(blocks)

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in unstack(self, unstacker_func)
   4337                 partial(unstacker_func,
   4338                         value_columns=self.items[blk.mgr_locs.indexer]),
-> 4339                 new_columns)
   4340 
   4341             new_blocks.extend(blocks)

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in _unstack(self, unstacker_func, new_columns)
   1808 
   1809         blocks = [self.make_block_same_class(vals, [place])
-> 1810                   for vals, place in zip(new_values, new_placement)]
   1811         return blocks, mask
   1812 

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in <listcomp>(.0)
   1808 
   1809         blocks = [self.make_block_same_class(vals, [place])
-> 1810                   for vals, place in zip(new_values, new_placement)]
   1811         return blocks, mask
   1812 

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in make_block_same_class(self, values, placement, fastpath, **kwargs)
    222             placement = self.mgr_locs
    223         return make_block(values, placement=placement, klass=self.__class__,
--> 224                           fastpath=fastpath, **kwargs)
    225 
    226     @mgr_locs.setter

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in make_block(values, placement, klass, ndim, dtype, fastpath)
   2948     elif klass is DatetimeTZBlock and not is_datetimetz(values):
   2949         return klass(values, ndim=ndim, fastpath=fastpath,
-> 2950                      placement=placement, dtype=dtype)
   2951 
   2952     return klass(values, ndim=ndim, fastpath=fastpath, placement=placement)

/usr/local/lib/python3.5/dist-packages/pandas/core/internals.py in __init__(self, values, placement, ndim, **kwargs)
   2610 
   2611         if values.tz is None:
-> 2612             raise ValueError("cannot create a DatetimeTZBlock without a tz")
   2613 
   2614         super(DatetimeTZBlock, self).__init__(values, placement=placement,

ValueError: cannot create a DatetimeTZBlock without a tz
```

#### Expected Output
```
   c timestamp
b  b         b
a             
a  c       2017-08-27 01:00:00.709949+00:00
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-1039-aws
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C.UTF-8
LANG: C.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.2.3
pip: 9.0.1
setuptools: 36.6.0
Cython: 0.27.2
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: 2.7.3.2 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### LINKED ISSUE #19420: DF.stack() on single column datetime with timezone causes loss of timezone
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

ts1 = list(pd.date_range(freq="D", start="20180101", end="20180103"))
df1 = pd.DataFrame({"A": ts1}, index=["a", "b", "c"])
print("Works as expected; no tz")
print(df1.stack())
#a  A   2018-01-01
#b  A   2018-01-02
#c  A   2018-01-03
#dtype: datetime64[ns]

ts2 = list(pd.date_range(freq="D", start="20180101", end="20180103", tz="America/New_York"))
ts3 = list(pd.date_range(freq="D", start="20180104", end="20180106", tz="America/New_York"))

df2 = pd.DataFrame({"A": ts2, "B": ts3}, index=["a", "b", "c"])
print("Works as expected; has tz")
print(df2.stack())
#a  A   2018-01-01 00:00:00-05:00
#   B   2018-01-04 00:00:00-05:00
#b  A   2018-01-02 00:00:00-05:00
#   B   2018-01-05 00:00:00-05:00
#c  A   2018-01-03 00:00:00-05:00
#   B   2018-01-06 00:00:00-05:00
#dtype: datetime64[ns, America/New_York]

df3 = pd.DataFrame({"A": ts2}, index=["a", "b", "c"])
print("Unexpected loss of tz, but converted to UTC")
print(df3.stack())
#a  A   2018-01-01 05:00:00
#b  A   2018-01-02 05:00:00
#c  A   2018-01-03 05:00:00
#dtype: datetime64[ns]

```
#### Problem description
Stacking a DF with multiple datetime with timezone columns seems fine but stacking a single datetime with timezone column appears to convert the time to a naive datetime.

This can be worked around using the .dt.tz and .dt.tz_localize/.dt.tz_convert functionality

#### Expected Output
a  A   2018-01-01 00:00:00-05:00
b  A   2018-01-02 00:00:00-05:00
c  A   2018-01-03 00:00:00-05:00
dtype: datetime64[ns, America/New_York]

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.21.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.utf-8
LANG: en_US.utf-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 28.8.0
Cython: None
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #21184: bug: merge_asof with tz-aware datetime "by" parameter raises
#### Code Sample
```python
import pandas as pd

left = pd.DataFrame({'by_col': pd.DatetimeIndex(['2018-01-01']).tz_localize('UTC'),
                     'on_col': [2], 'values': ['a']})
right = pd.DataFrame({'by_col': pd.DatetimeIndex(['2018-01-01']).tz_localize('UTC'),
                      'on_col': [1], 'values': ['b']})

merged = pd.merge_asof(left, right, by='by_col', on='on_col')
```

#### Error traceback
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Hamb\AppData\Local\Programs\Python\Python36\lib\site-packages\pandas\core\reshape\merge.py", line 478, in merge_asof
    return op.get_result()
  File "C:\Users\Hamb\AppData\Local\Programs\Python\Python36\lib\site-packages\pandas\core\reshape\merge.py", line 1163, in get_result
    join_index, left_indexer, right_indexer = self._get_join_info()
  File "C:\Users\Hamb\AppData\Local\Programs\Python\Python36\lib\site-packages\pandas\core\reshape\merge.py", line 776, in _get_join_info
    right_indexer) = self._get_join_indexers()
  File "C:\Users\Hamb\AppData\Local\Programs\Python\Python36\lib\site-packages\pandas\core\reshape\merge.py", line 1437, in _get_join_indexers
    tolerance)
TypeError: Argument 'left_by_values' has incorrect type (expected numpy.ndarray, got Index)
```

#### Problem description
Function pandas.merge_asof raises when "by" parameter is provided a column of tz-aware datetime type.
Note that the same code with tz-naive datetimes works :
```python
import pandas as pd

left = pd.DataFrame({'by_col': pd.DatetimeIndex(['2018-01-01']),
                     'on_col': [2], 'values': ['a']})
right = pd.DataFrame({'by_col': pd.DatetimeIndex(['2018-01-01']),
                      'on_col': [1], 'values': ['b']})

merged = pd.merge_asof(left, right, by='by_col', on='on_col')
print(merged)
```
outputs :
```
    by_col  on_col values_x values_y
0 2018-01-01       2        a        b
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: FR
LOCALE: None.None
pandas: 0.23.0
pytest: None
pip: 10.0.1
setuptools: 38.5.1
Cython: None
numpy: 1.14.1
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #22905: BUG: to_period behaves different between Timestamp and DatetimeIndex with timezones
After 935c5d145 `pandas/tests/indexes/datetimes/test_astype.py` fails depending on the timezone:
in fact 
```
TZ=Etc/GMT+1 pytest pandas/tests/indexes/datetimes/test_astype.py -k TestToPeriod
```
is OK while
```
TZ=Etc/GMT-1 pytest pandas/tests/indexes/datetimes/test_astype.py -k TestToPeriod
```
fails.

This can be nailed down to the following code sample.

#### Code Sample

```python
>>> ts = pd.date_range('1/1/2000', '2/1/2000', tz='Etc/GMT-1')
>>> ts.to_period()[0] == ts[0].to_period()
False
>>> ts.to_period()[0], ts[0].to_period()
(Period('1999-12-31', 'D'), Period('2000-01-01', 'D'))
>>> ts = pd.date_range('1/1/2000', '2/1/2000', tz='Etc/GMT+1')
>>> ts.to_period()[0] == ts[0].to_period()
True
>>> ts.to_period()[0], ts[0].to_period()
(Period('2000-01-01', 'D'), Period('2000-01-01', 'D'))
```
#### Problem description

It seems that `ts.to_period()[0] == ts[0].to_period()` depends on tzinfo, true if west of GMT, false if east of GMT. Before 935c5d145 `ts.to_period()[0] == ts[0].to_period()` is always `True` irrespective of tzinfo.

#### Output of ``pd.show_versions()``

<details>

<pre>
INSTALLED VERSIONS
------------------
commit: 935c5d1453938a284805cb7d3c0c3255390997a7
python: 3.7.0.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0.dev0+555.g935c5d145
pytest: 3.7.4
pip: 18.0
setuptools: 39.0.1
Cython: 0.28.5
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: 0.10.9
IPython: 6.5.0
sphinx: 1.7.9
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: 1.5.1
bottleneck: 1.2.1
tables: None
numexpr: 2.6.8
feather: None
matplotlib: 3.0.0
openpyxl: 2.5.5
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.1.1
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.12
pymysql: 0.9.2
psycopg2: None
jinja2: 2.10
s3fs: 0.1.6
fastparquet: 0.1.6
pandas_gbq: None
pandas_datareader: None
gcsfs: 0.1.2
</pre>

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #22905
- [x] closes #18338
- [x] closes #19420
- [x] closes #21184
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_reshape.py [hunk 1]
```diff
@@ -936,3 +936,36 @@ def test_unstack_fill_frame_object():
...
+
+def test_unstack_timezone_aware_values():
+    # GH 18338
+    df = pd.DataFrame({
+        'timestamp': [
+            pd.Timestamp('2017-08-27 01:00:00.709949+0000', tz='UTC')],
+        'a': ['a'],
+        'b': ['b'],
+        'c': ['c'],
+    })
```
Comment (TomAugspurger) [inline pandas/tests/frame/test_reshape.py]: looks like this may need a `columns=['timestamp', 'a', 'b', 'c']` to ensure the right order on Py2.

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ex @TomAugspurger comment, ping on green.

**Comment by jreback** [comment]
@mroeschke can you merge master

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**